### PR TITLE
[5.1] Update ProviderRepository::writeManifest

### PR DESCRIPTION
**Pull request by CupOfTea696**
Make sure `ProviderRepository::writeManifest` always returns an array that has the `when` key.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Maybe this is better to go into 5.2? Ping @taylorotwell.

**Comment by crynobone** [comment]
In what situation does the `when` key missing?

**Comment by CupOfTea696** [comment]
@crynobone In a Package I'm building, I'm manually creating another ProviderRepository to load additional ServiceProviders from a ServiceProvider's boot method. I'm doing this to dynamically load ServiceProviders that are marked as "active" in the Database.

**Comment by CupOfTea696** [comment]
@GrahamCampbell @taylorotwell well, from what I see it's definitely not a breaking change so I don't see why it can't go in 5.1. But the decision is yours of course :)

**Comment by crynobone** [comment]
If you're loading additional service provider using `$this->app->register()`, does additional service provider will be loaded eagerly and there no reason to change `when`.

**Comment by CupOfTea696** [comment]
I'm using the following method I added to my ServiceProvider:

```php
protected function loadServiceProviders($providers)
    {
        $providers = is_string($providers) ? [$providers] : $providers;
        
        $manifestPath = app()->getCachedServicesPath();
        (new ProviderRepository($this->app, new Filesystem, $manifestPath))
                    ->load($providers);
    }
```

Loading additional ServiceProviders this way does cause an error saying the `when` index is undefined.

**Comment by crynobone** [comment]
> well, from what I see it's definitely not a breaking change so I don't see why it can't go in 5.1

It also has to make sense, because right now you're only merging it on return, but the line before which write the content to `services.json` doesn't include this change.

----

> I'm using the following method I added to my ServiceProvider

Why would you do this? loading `ProviderRepository` twice in a single request would only double the work needed to resolve service provider, and since you're running it twice, the `service.json` will be overwritten twice on each request.

`services.json` was meant to allow service providers resolution to be cached, and only change when needed.

**Comment by CupOfTea696** [comment]
@crynobone Still working on this, already updated the services.json to an additional cached file.
Loading the ProviderRepository is indeed extra work, but since it was never resolved in the service container it's my only option. If I use the Application's Register method, it will not take deferred services into account.

**Comment by crynobone** [comment]
I also build database driven service provider (for extension/plugin development) and I can assure you this is not the way to do it. Take a look at <https://github.com/orchestral/extension/blob/master/src/ProviderRepository.php> on how to resolve this additional service providers.

**Comment by CupOfTea696** [comment]
@crynobone besides the different cache file (which I have already changed by now) I don't see how your method is that much different from loading the `ProviderRepository` again. It just seems like a lot of code duplication to me.

**Comment by CupOfTea696** [comment]
@crynobone 
> It also has to make sense, because right now you're only merging it on return, but the line before which write the content to services.json doesn't include this change.

That's because there is no need to write the empty when array to the cache file. If you scroll up you can see the the `loadManifest` method already automatically adds the `when` key to the array. So saving it would only make the file slightly larger.

**Comment by crynobone** [comment]
> I don't see how your method is that much different from loading the ProviderRepository again. It just seems like a lot of code duplication to me.

1. When Laravel first load, during `Illuminate\Foundation\Bootstrap\RegisterProviders` Laravel would load the manifest file for the request, let assume that it's empty, it has to construct every service providers to detect 1) is it eager/deferred 2) does it loaded by event (using when) 2) which service does it provider (if deferred) https://github.com/laravel/framework/blob/5.1/src/Illuminate/Foundation/ProviderRepository.php#L106-L136. Thie result would then be cached so that the next request doesn't have to go through this scanning process.

2. During `Illuminate\Foundation\Bootstrap\BootProviders`, you construct a `ProviderRepository` for the second time with a different set of `$providers`. This time manifest file content is available, however since you are sending a different service provider `$this->shouldRecompile($manifest, $providers)` would return false, and it need to scan the service providers again, save it to `services.json`.

3. Then you make a second request, Laravel goes back to `app.providers` and found that the list is different from (2), and scanning happens again.

> It just seems like a lot of code duplication to me.

The purpose of `services.json` is that service providers scanning only happen when there a change to the list, otherwise you just constructing all the service providers all the time, regardless whether it's deferred or eagered.

**Comment by CupOfTea696** [comment]
@crynobone as I said twice before now, I have changed the services.json file. It now uses a different file name using `{{vendor}}_services.json`. 

screen shot 2015-10-20 at 10 35 15 am
screen shot 2015-10-20 at 10 35 28 am

**Comment by crynobone** [comment]
> as I said twice before now, I have changed the services.json file. It now uses a different file name using {{vendor}}_services.json.

If you already done this, do you still need this PR?

**Comment by CupOfTea696** [comment]
@crynobone yes, because I'm still using Laravel's `ProviderRepository` class to load my ServiceProviders, I just pass in a different manifest file name when I create it. I still get the error that the `when` key isn't set.

File: src/Illuminate/Foundation/ProviderRepository.php [hunk 1]
```diff
@@ -74,7 +74,7 @@ public function load(array $providers)
             $this->app->register($this->createProvider($provider));
         }
 
-        $this->app->setDeferredServices($manifest['deferred']);
+        $this->app->addDeferredServices($manifest['deferred']);
```
Comment (taylorotwell) [inline src/Illuminate/Foundation/ProviderRepository.php]: Why is this change necessary?

Comment (CupOfTea696) [inline src/Illuminate/Foundation/ProviderRepository.php]: When loading extra service providers using a ProviderRepository later in the Application, currently it would replace all deferred services that were already registered, rather than just adding additional deferred services.
